microceph/ceph/replication.go

```diff
@@ -0,0 +1,127 @@
+	"github.com/canonical/microceph/microceph/constants"
+	"github.com/canonical/microceph/microceph/interfaces"
+	"github.com/qmuntal/stateless"
+)
+
+type repArgIndex int
+type ReplicationState string
+
+const (
+	StateDisabledReplication ReplicationState = "replication_disabled"
+	StateEnabledReplication  ReplicationState = "replication_enabled"
+)
+
+const (
+	repArgHandler  repArgIndex = 0
+	repArgResponse repArgIndex = 1
+	repArgState    repArgIndex = 2
+)
+
+type ReplicationHandlerInterface interface {
+	PreFill(ctx context.Context, request types.ReplicationRequest) error
+	GetResourceState() ReplicationState
+	EnableHandler(ctx context.Context, args ...any) error
+	DisableHandler(ctx context.Context, args ...any) error
+	ConfigureHandler(ctx context.Context, args ...any) error
+	ListHandler(ctx context.Context, args ...any) error
+	StatusHandler(ctx context.Context, args ...any) error
+}
+
+func GetReplicationHandler(name string) ReplicationHandlerInterface {
+	// Add RGW and CephFs Replication handlers here.
+	table := map[string]ReplicationHandlerInterface{
+		"rbd": &RbdReplicationHandler{},
+	}
+
+	rh, ok := table[name]
+	if !ok {
+		return nil
+	}
+
+	return rh
+}
+
+func GetReplicationStateMachine(initialState ReplicationState) *stateless.StateMachine {
+	newFsm := stateless.NewStateMachine(initialState)
+	// Configure transitions for disabled state.
+	newFsm.Configure(StateDisabledReplication).
+		Permit(constants.EventEnableReplication, StateEnabledReplication).
+		OnEntryFrom(constants.EventDisableReplication, disableHandler).
+		InternalTransition(constants.EventListReplication, listHandler).
+		InternalTransition(constants.EventDisableReplication, disableHandler)
+
+	// Configure transitions for enabled state.
+	newFsm.Configure(StateEnabledReplication).
+		Permit(constants.EventDisableReplication, StateDisabledReplication).
+		OnEntryFrom(constants.EventEnableReplication, enableHandler).
+		InternalTransition(constants.EventEnableReplication, enableHandler).
+		InternalTransition(constants.EventConfigureReplication, configureHandler).
+		InternalTransition(constants.EventListReplication, listHandler).
+		InternalTransition(constants.EventStatusReplication, statusHandler)
+
+	// Check Event params type.
+	var output *string
+	var dummyState interfaces.CephState
+	var eventHandler ReplicationHandlerInterface
+	inputType := reflect.TypeOf(&eventHandler).Elem()
+	outputType := reflect.TypeOf(output)
+	stateType := reflect.TypeOf(dummyState)
+	newFsm.SetTriggerParameters(constants.EventEnableReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventDisableReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventConfigureReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventListReplication, inputType, outputType, stateType)
+	newFsm.SetTriggerParameters(constants.EventStatusReplication, inputType, outputType, stateType)
```

Comment:
Nit, maybe less repetitive sth like

```
   triggers := []stateless.Trigger{
        constants.EventEnableReplication,
        constants.EventDisableReplication,
        constants.EventConfigureReplication,
        constants.EventListReplication,
        constants.EventStatusReplication,
    }

    for _, trigger := range triggers { 
newFsm.SetTriggerParameters(trigger, inputType, outputType, stateType)
 }
```

Reply:
ack, I have reformatted this to a condensed version.